docs/rfcs/GPUXDialect.md

```diff
@@ -0,0 +1,226 @@
+
+
+#### gpux.launch_func (gpux::LaunchFuncOp)
+
+Launches a function as a GPU kernel.
+
+Syntax:
+
+```
+operation ::= gpux.launch_func custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              $kernel
+              blocks in  ($gridSizeX, $gridSizeY, $gridSizeZ)
+              threads in ($blockSizeX, $blockSizeY, $blockSizeZ)
+              (dynamic_shared_memory_size $dynamicSharedMemorySize^)?
+              custom<LaunchFuncOperands>($gpux_stream, $operands, type($operands)) attr-dict
+```
+
+This op is an extension of upstream gpu.launch_func op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+#### gpux.wait (gpux::WaitOp)
+
+Wait for gpu ops in a particular stream to complete.
+
+Syntax:
+
+```
+operation ::= gpu.wait custom<AsyncDependencies>(type($asyncToken), $asyncDependencies)
+              $gpux_stream attr-dict
+```
+
+The wait op will need to be provided with the stream to wait on. This op is an extension of upstream gpu.wait op with one added argument for stream. Stream here will be an optional argument, so when upstreaming the code, the upstream GPU dialect users have no change in their usage of this op.
+
+## Example Usage in case of user provided stream:
+
+// IR before our custom transformation pass:
+
+```
+
+func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  %cst = arith.constant 2.200000e+00 : f32
+  %cst_0 = arith.constant 1.100000e+00 : f32
+  %cst_1 = arith.constant 0.000000e+00 : f32
+  %0 = gpu.alloc  () : memref<8xf32>
+  %1 = gpu.alloc  () : memref<8xf32>
+  %2 = gpu.alloc  () : memref<8xf32>
+  call @fillResource1DFloat(%0, %cst_0) : (memref<?xf32>, f32) -> ()
+  call @fillResource1DFloat(%1, %cst) : (memref<?xf32>, f32) -> ()
+  call @fillResource1DFloat(%2, %cst_1) : (memref<?xf32>, f32) -> ()
+  gpu.launch_func  @main_kernel::@main_kernel blocks in (%c8, %c1, %c1) threads in (%c1, %c1, %c1) args(%0 : memref<8xf32>, %1 : memref<8xf32>, %2 : memref<8xf32>)
+  %3 = memref.cast %2 : memref<8xf32> to memref<*xf32>
+  call @printMemrefF32(%3) : (memref<*xf32>) -> ()
+  gpu.dealloc(%0) : memref<8xf32>
+  gpu.dealloc(%1) : memref<8xf32>
+  gpu.dealloc(%2) : memref<8xf32>
+  return
+}
+
+```
+
+// IR after our custom transformation pass:
+
+```
+
+func.func @main(%stream : !gpux.StreamType) attributes {llvm.emit_c_interface} {
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  %cst = arith.constant 2.200000e+00 : f32
+  %cst_0 = arith.constant 1.100000e+00 : f32
+  %cst_1 = arith.constant 0.000000e+00 : f32
+  %0 = gpux.alloc (%stream) : (!gpux.StreamType) -> memref<8xf32>
```

Comment:
It doesn't make sense. How the optimization pass would know that it needs tot take the stream passed by the function and assign it to this operations?  What about user pass two streams in the main function parameters? What if the operations within  program itself needs to two streams?  There should be a separate optimization pass to decide which stream to use, and that optimization pass need to decide which op to use which stream. Not just pass it as parameter, hoping another optimization pass to fix the rest.

Reply:
@Jianhui-Li That's the part which do not understand. The mere existence of a stream-creation operation somewhere in the code give no hint about where to use it. Hence we suggested using regions (see #283). Can you explain how an analysis would be able to do this mapping? What's the contract for linking create_stream to the alloc?